pyethapp/rpc_client.py

```diff
@@ -538,55 +661,81 @@ def poll(self, transaction_hash, confirmations=None, timeout=None):
 
         if len(transaction_hash) != 32:
-            raise ValueError('transaction_hash length must be 32 (it might be hex encode)')
+            raise ValueError(
+                'transaction_hash length must be 32 (it might be hex encode)'
+            )
+
+        transaction_hash = data_encoder(transaction_hash)
 
         deadline = None
         if timeout:
-            deadline = time.time() + timeout
+            deadline = gevent.Timeout(timeout)
+            deadline.start()
 
-        transaction_hash = data_encoder(transaction_hash)
+        try:
+            # used to check if the transaction was removed, this could happen
+            # if gas price is to low:
+            #
+            # > Transaction (acbca3d6) below gas price (tx=1 Wei ask=18
+            # > Shannon). All sequential txs from this address(7d0eae79)
+            # > will be ignored
+            #
+            last_result = None
+
+            while True:
+                # Could return None for a short period of time, until the
+                # transaction is added to the pool
+                transaction = self.call('eth_getTransactionByHash', transaction_hash)
+
+                # if the transaction was added to the pool and then removed
+                if transaction is None and last_result is not None:
+                    raise Exception('invalid transaction, check gas price')
+
+                # the transaction was added to the pool and mined
+                if transaction and transaction['blockNumber'] is not None:
+                    break
 
-        transaction = self.call('eth_getTransactionByHash', transaction_hash)
-        while transaction is None or transaction["blockNumber"] is None:
-            if deadline and time.time() > deadline:
-                raise Exception('timeout when polling for transaction')
+                last_result = transaction
 
-            gevent.sleep(.5)
-            transaction = self.call('eth_getTransactionByHash', transaction_hash)
+                gevent.sleep(.5)
 
-        if confirmations is None:
-            return
+            if confirmations:
+                # this will wait for both APPLIED and REVERTED transactions
+                transaction_block = quantity_decoder(transaction['blockNumber'])
+                confirmation_block = transaction_block + confirmations
 
-        # this will wait for both APPLIED and REVERTED transactions
-        transaction_block = quantity_decoder(transaction['blockNumber'])
-        confirmation_block = transaction_block + confirmations
+                block_number = self.blocknumber()
 
-        block_number = self.blocknumber()
-        while confirmation_block > block_number:
-            if deadline and time.time() > deadline:
-                raise Exception('timeout when waiting for confirmation')
+                while block_number < confirmation_block:
+                    gevent.sleep(.5)
+                    block_number = self.blocknumber()
 
```

Comment:
Now returning the current block number, this will include the confirmations block.